### PR TITLE
Batch Argus Annotation uid retrieval and add caching

### PR DESCRIPTION
**Pull request by naveenreddykarri**
Needs more testing. Will cleanup the added debug statements.
Added unit tests. It does batching on put metrics for annotations, get is still not batched and then final add annotations which is already batched. It also maintains a LRU cache for uids that are retrieved.

### CODE REVIEW

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java [hunk 1]
```diff
@@ -128,6 +132,8 @@
 	protected final MonitorService _monitorService;
 	private final int RETRY_COUNT;
 
+	private Cache<String, String> _annotationUidCache;
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java]: Can you add a java comment what the String key,value in map correspond to

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java [hunk 2]
```diff
@@ -386,18 +398,109 @@ public void putMetrics(List<Metric> metrics) {
...
+
+				if(StringUtils.isEmpty(uid)) {
+					keyAnnotationMap.put(key, annotation);
+				} else {
+					AnnotationWrapper wrapper = new AnnotationWrapper(uid, annotation);
+					wrappers.add(wrapper);
+				}
+			}
+
+			Map<String, String> keyUidMap = getUids(keyAnnotationMap);
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java]: Can you add comment
If there is a local cache miss then the annotation uids that should be fetched from tsdb is in keyAnnotationMap

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java [hunk 3]
```diff
@@ -386,18 +398,109 @@ public void putMetrics(List<Metric> metrics) {
...
 				_retry(wrappers, _roundRobinIterator);
 			}
 		}
 	}
 
+	private Map<String, String> getUids(Map<String, Annotation> annotationMap) {
+
+		List<MetricQuery> queries = new ArrayList<>();
+
+		List<Metric> metrics = new ArrayList<>();
```
Comment (dilipdevaraj-sfdc) [inline ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java]: remove extra lines

**Comment by dilipdevaraj-sfdc** [comment]
something to look at
https://github.com/OpenTSDB/opentsdb/issues/913

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java (hunk 1 above)
Comment (naveenreddykarri) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java]: fixed

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java (hunk 2 above)
Comment (naveenreddykarri) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java]: fixed

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java (hunk 3 above)
Comment (naveenreddykarri) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java]: fixed

File: ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java [hunk 4]
```diff
@@ -386,18 +398,109 @@ public void putMetrics(List<Metric> metrics) {
...
+			putMetrics(metrics);
+			try {
+				Thread.sleep(backOff);
+			} catch (InterruptedException ex) {
+				// We continue if interrupted.
+			}
+
+			try {
+
+				Map<String, String> uids = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (naveenreddykarri) [reply ArgusCore/src/main/java/com/salesforce/dva/argus/service/tsdb/AbstractTSDBService.java]: fixed

**Pull request merged**